paddle/fluid/operators/fill_diagonal_tensor_op.cu

```diff
@@ -0,0 +1,204 @@
+    auto sumoff = matdim[i] + offset;
+    for (int j = threadIdx.y; j < fill_dims1; j += blockDim.y) {
+      auto fill_index = j * (strides[1] + strides[0]) + sumoff;
+      if (fill_index < size) {
+        dx_data[fill_index] = T(0);
+      }
+    }
+  }
+}
+
+template <typename T>
+class FillDiagonalTensorCUDAKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext &ctx) const override {
+#ifdef __HIPCC__
+    const int64_t kMaxBlockDim = 256;
+#else
+    const int64_t kMaxBlockDim = 512;
+#endif
+    auto *out = ctx.Output<framework::Tensor>("Out");
+    auto *srctensor = ctx.Input<framework::Tensor>("src_tensor");
+    auto dim1 = ctx.Attr<int>("dim1");
+    auto dim2 = ctx.Attr<int>("dim2");
+    auto offset = ctx.Attr<int>("offset");
+
+    auto *xin = ctx.Input<framework::Tensor>("X");
+    framework::TensorCopy(*xin, ctx.GetPlace(), out);
+
+    T *out_data = out->mutable_data<T>(ctx.GetPlace());
+    T *fill_data = const_cast<T *>(srctensor->data<T>());
+
+    auto out_dims = out->dims();
+    auto matdims = srctensor->dims();
+    auto fill_dims = flatten_to_2d(matdims, matdims.size() - 1);
+
+    int new_dims[2];
+    int *memory_block = new int[2 + fill_dims[0]]{0};
+    int *strides = &memory_block[0];
+    int *matdim = &memory_block[2];
+    CalMatDims(out_dims, dim1, dim2, &offset, new_dims, strides, matdim);
+    PADDLE_ENFORCE_EQ(
+        new_dims[0], fill_dims[0],
+        platform::errors::InvalidArgument("The dims should be %d x %d, but get "
+                                          "%d x %d in src_tensor filled in",
+                                          new_dims[0], new_dims[1],
+                                          fill_dims[0], fill_dims[1]));
+    PADDLE_ENFORCE_EQ(
+        new_dims[1], fill_dims[1],
+        platform::errors::InvalidArgument("The dims should be %d x %d, but get "
+                                          "%d x %d in src_tensor filled in",
+                                          new_dims[0], new_dims[1],
+                                          fill_dims[0], fill_dims[1]));
+
+    auto size = out->numel();
+    int *memory_block_cu;
+#ifdef __HIPCC__
+    hipMalloc(reinterpret_cast<void **>(&memory_block_cu),
+              sizeof(int) * (2 + fill_dims[0]));
+    hipMemcpy(memory_block_cu, memory_block, sizeof(int) * (2 + fill_dims[0]),
+              hipMemcpyHostToDevice);
+#else
+    cudaMalloc(reinterpret_cast<void **>(&memory_block_cu),
+               sizeof(int) * (2 + fill_dims[0]));
+    cudaMemcpy(memory_block_cu, memory_block, sizeof(int) * (2 + fill_dims[0]),
+               cudaMemcpyHostToDevice);
+#endif
+    int *strides_cu = &memory_block_cu[0], *matdim_cu = &memory_block_cu[2];
+
+    dim3 kBlockDim(std::min(int64_t(new_dims[0]), kMaxBlockDim),
+                   std::min(int64_t(new_dims[1]), kMaxBlockDim), 1);
+    fill_diagonal_tensor_kernel<T><<<1, kBlockDim, 0>>>(
```

Comment:
I do not know why you use this dimension setting. I prefer to use both `gridDim` and `blockDim`. The `gridDim` can be up to `2^27` usually.

Also, add large data testing.

Reply:
done